library_management_system_downloader/create_api_request.py

```diff
@@ -0,0 +1,51 @@
+import backoff
+from ratelimit import rate_limited
+import requests
+import logging
+
+
+class ErrorWithAPI(Exception):
+    pass
+
+
+# From the backoff documentation (https://pypi.python.org/pypi/backoff),
+# set maximum number of tries on a failed download:
+@backoff.on_exception(
+        backoff.expo,
+        requests.exceptions.RequestException,
+        max_tries=8)
+# From the ratelimit documentation (https://pypi.python.org/pypi/ratelimit),
+# set the rate limit (in seconds) for API calls:
+@rate_limited(0.2)
+def create_api_request(
+        item_doi,
+        api_base_url,
+        static_api_request_parameters_dictionary=None,
+        custom_user_agent_string=None):
+    """Given an item DOI (lacking the 'doi:/' prefix), query an OpenURL
+    resolver) and return XML from it."""
+
+    if static_api_request_parameters_dictionary is None:
+        static_api_request_parameters_dictionary = {}
+
+    custom_api_query_header = {}
+    if custom_user_agent_string:
+        custom_api_query_header['user-agent'] = custom_user_agent_string
+
+    # Update the static api parameters to include the (dynamic) DOI:
+    api_request_parameters = static_api_request_parameters_dictionary.copy()
+    api_request_parameters.update({
+            'rft_id': f'info:doi/{item_doi}'})
+
+    api_response = requests.get(
+            api_base_url,
+            params=api_request_parameters,
+            headers=custom_api_query_header)
+
+    if api_response.status_code != 200:
+        raise ErrorWithAPI('Problem contacting API: {}'.format(
+                api_response.status_code))
```

Comment:
Perhaps log the actual `api_response.text()` under `info` or `warning`? This may contain why the query failed.

Reply:
We don't want non-200 responses to crash the program. They should perhaps be logged as warnings? In other words we need to know about these failures, but the querying should continue. Do you plan on handling these errors anywhere?

Reply:
Do you mean to replace `raise ErrorWithAPI` with, e.g., `logging.warning()`? I'm fine with that.

I didn't plan on handling them elsewhere for now, as I actually thought it would be desirable for it to crash the program (since it would have up to 8 tries via `backoff`, a non-200 response would presumably mean more than a momentary connectivity interruption). The API isn't well-documented enough that I know what status codes might be returned, so I figured that if there's a problem, I should take a look at it manually, at least to start.

I've added the full response text (and switched to f-strings in this code chunk) in 658f7b4. If you would still like to see it not crash, though, let me know here, and I'd be fine to switch it to `logging.warning()`.